test/integration/suite_test.go

```diff
@@ -309,3 +313,47 @@ func waitForBpfdConfigDelete(ctx context.Context, env environments.Environment)
 		}
 	}
 }
+
+func waitForDataplaneReadiness(ctx context.Context, env environments.Environment) error {
+	for {
+		time.Sleep(2 * time.Second)
+		select {
+		case <-ctx.Done():
+			if err := ctx.Err(); err != nil {
+				return fmt.Errorf("context completed while waiting for dataplane ready: %w", err)
+			}
+			return fmt.Errorf("context completed while waiting for dataplane ready")
+		default:
+			fmt.Println("INFO: waiting for dataplane ready")
+
+			dataplanes, err := env.Cluster().Client().CoreV1().Pods(vars.DefaultNamespace).
+				List(ctx, metav1.ListOptions{LabelSelector: fmt.Sprintf("component=%s", vars.DefaultDataPlaneComponentLabel)})
+			if err != nil {
+				return fmt.Errorf("failed to fetch dataplane pod list: %w", err)
+			}
+
+			// Check whether dataplane is set readiness probe.
+			for _, container := range dataplanes.Items[0].Spec.Containers {
+				if container.Name == "dataplane" && container.ReadinessProbe == nil {
+					return fmt.Errorf("container doesn't have readiness probe")
+				}
+			}
+
+			// Check all pods' readiness
+			var ready int
+			for _, pod := range dataplanes.Items {
+				for _, status := range pod.Status.ContainerStatuses {
+					if status.Name == "dataplane" && status.Ready {
+						ready++
+					}
+				}
+			}
+
+			if ready != len(dataplanes.Items) {
+				return fmt.Errorf("dataplane not ready")
```

Comment:
We have two returns above with errors that indicate problems we don't expect to have while waiting for the dataplane (API connectivity issues and missing readiness probes) which are both good checks.

However _this_ return creates an early exit and basically means that the `for` loop which encompasses it is not practically exercised, we could break here instead of returning an error:

```suggestion
				fmt.Printf("%d dataplanes not yet ready", len(dataplanes.Items) - ready)
				time.Sleep(time.Second) // small rest from hitting the API over and over again
				break
```

This will make sure the `for` loop takes effect and retries the listing when the pods are not yet ready. Then too we can put the `time.Sleep` here so that we only pause if needed to between API calls.

Reply:
Nice, I misunderstood exitonError's behaviour